packages/parse5-sax-parser/lib/index.ts

```diff
@@ -0,0 +1,312 @@
+        this.lastChunkWritten = true;
+        callback(null, this._transformChunk(''));
+    }
+
+    /**
+     * Stops parsing. Useful if you want the parser to stop consuming CPU time
+     * once you've obtained the desired info from the input stream. Doesn't
+     * prevent piping, so that data will flow through the parser as usual.
+     *
+     * @example
+     *
+     * ```js
+     * const SAXParser = require('parse5-sax-parser');
+     * const http = require('http');
+     * const fs = require('fs');
+     *
+     * const file = fs.createWriteStream('google.com.html');
+     * const parser = new SAXParser();
+     *
+     * parser.on('doctype', ({ name, publicId, systemId }) => {
+     *     // Process doctype info and stop parsing
+     *     ...
+     *     parser.stop();
+     * });
+     *
+     * http.get('http://google.com', res => {
+     *     // Despite the fact that parser.stop() was called whole
+     *     // content of the page will be written to the file
+     *     res.pipe(parser).pipe(file);
+     * });
+     * ```
+     */
+    public stop(): void {
+        this.stopped = true;
+    }
+
+    //Internals
+    protected _transformChunk(chunk: string): string {
+        if (!this.stopped) {
+            this.tokenizer.write(chunk, this.lastChunkWritten);
+            this._runParsingLoop();
+        }
+        return chunk;
+    }
+
+    private _runParsingLoop(): void {
+        let token = null;
+
+        do {
+            token = this.parserFeedbackSimulator.getNextToken();
+
+            if (token.type === TokenType.HIBERNATION) {
+                break;
+            }
+
+            if (
+                token.type === TokenType.CHARACTER ||
+                token.type === TokenType.WHITESPACE_CHARACTER ||
+                token.type === TokenType.NULL_CHARACTER
+            ) {
+                if (this.pendingText === null) {
+                    token.type = TokenType.CHARACTER;
+                    this.pendingText = token;
+                } else {
+                    this.pendingText.chars += token.chars;
+
+                    if (token.location && this.pendingText.location) {
+                        const { endLine, endCol, endOffset } = token.location;
+                        this.pendingText.location = {
+                            ...this.pendingText.location,
+                            endLine,
+                            endCol,
+                            endOffset,
+                        };
+                    }
+                }
+            } else {
+                this._emitPendingText();
+                this._handleToken(token);
+            }
+        } while (!this.stopped && token.type !== TokenType.EOF);
+    }
+
+    protected _handleToken(token: Token): boolean {
+        if (token.type === TokenType.EOF) {
+            return true;
+        }
+
+        const { eventName, reshapeToken } = TOKEN_EMISSION_HELPERS[token.type];
```

Comment:
this pattern (`TOKEN_EMISSION_HELPERS`) will always result in a disconnect of our types, i think. i had a think and im not sure there's a way to strongly type the value of a record based on its key.

in future i think we should probably change this to a switch so we gain the type inference from it. something as simple as:

```ts
switch (token.type) {
  case TokenType.START_TAG:
    this._emitToken('startTag', reshapeStartTagToken(token));
    break;
  // ...
}
```

would be more readable IMO too

Reply:
Good callout, updated with a switch.